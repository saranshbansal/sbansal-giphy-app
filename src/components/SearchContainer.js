import React, { Component } from "react";
import { fetchByKeyword } from "../resources/api";
import GifList from "./GifList";
import Loader from "./library/Loader";
import Search from "./Search";

class SearchContainer extends Component {
  constructor(props) {
    super(props);

    this.handleSearch = this.handleSearch.bind(this);
    this.refreshResults = this.refreshResults.bind(this);
  }

  state = {
    gifs: [],
    limit: 20,
    offset: 0,
    loading: false,
    searchText: ""
  };

  async handleSearch(searchText, offset = 0) {
    const { limit } = this.state;

    // Always reset result on fresh search
    let gifsContent = [];

    try {
      this.setState({
        loading: true
      });

      const result = await fetchByKeyword(searchText, offset, limit);

      if (!!result && result.data) {
        result.data.map(gif => {
          return gifsContent.push({
            id: gif.id,
            imageUrl: gif.images.fixed_height_downsampled.url,
            staticImageUrl: gif.images.fixed_height_still.url,
            title: gif.title,
          });
        });
      }
    } catch (e) {
      console.error(e);
    } finally {
      this.setState({
        gifs: gifsContent,
        offset: offset,
        loading: false,
        searchText
      });
    }
  }

  async refreshResults() {
    const { gifs, searchText, limit, offset } = this.state;

    const nextOffset = offset + limit + 1;

    // Always append result on clicking `more...`
    let gifsContent = gifs;

    try {
      const result = await fetchByKeyword(searchText, nextOffset, limit);

      if (!!result && result.data) {
        result.data.map(gif => {
          return gifsContent.push({
            id: gif.id,
            imageUrl: gif.images.fixed_height_downsampled.url,
            staticImageUrl: gif.images.fixed_height_still.url,
            title: gif.title,
          });
        });
      }
    } catch (e) {
      console.error(e);
    } finally {
      this.setState({
        gifs: gifsContent,
        offset: nextOffset,
        loading: false
      });
    }
  }

  render() {
    const { loading, gifs } = this.state;

    const hasResults = !!gifs && gifs.length > 0;
    const showResults = hasResults && !loading;

    return (
      <div className="search">
        <Search onSearch={this.handleSearch} />

        {loading && <Loader />}

        {showResults && (
          <div className="gifs">
            <GifList data={gifs} />

            <button className="btn btn-more" onClick={this.refreshResults} aria-label="Show More">
              More...
            </button>
          </div>
        )}
      </div>
    );
  }
}

export default SearchContainer;
