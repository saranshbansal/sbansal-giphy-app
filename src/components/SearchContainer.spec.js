import { mount, shallow } from "enzyme";
import React from "react";
import SearchContainer from "./SearchContainer";

describe("SearchContainer", () => {
  it("should render correctly.", () => {
    const wrapper = shallow(<SearchContainer />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should render search box", () => {
    const wrapper = mount(<SearchContainer />);

    expect(wrapper.find(".search-box")).toHaveLength(1);
    expect(wrapper.state("loading")).toBe(false);
  });

  it("should render results against some keyword", async () => {
    const wrapper = mount(<SearchContainer />);
    const instance = wrapper.instance();
    
    expect(wrapper.state().gifs.length).toBe(0);
    
    await instance.handleSearch("random");

    expect(wrapper.state().gifs.length).toBeGreaterThan(0);
  });
});
