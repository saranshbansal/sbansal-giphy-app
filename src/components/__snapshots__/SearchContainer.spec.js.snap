// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchContainer should render correctly. 1`] = `
<div
  className="search"
>
  <Search
    onSearch={[Function]}
  />
</div>
`;
